src/call-api.test.ts

```diff
@@ -0,0 +1,394 @@
+      parameters: {
+        to: "USD",
+        from: "ETH",
+      },
+      apiCredentials: [
+        {
+          securitySchemeName,
+          securitySchemeValue,
+        },
+      ],
+      metadata: null,
+    });
+  });
+
+  it("returns null if templateId fails verification", async () => {
+    const spy = jest.spyOn(adapter, "buildAndExecuteRequest") as any;
+
+    const [logs, res] = await readApiValue(
+      airnodeAddress,
+      oises,
+      apiCredentials,
+      {
+        ...job,
+        templateParameters: [
+          ...job.templateParameters,
+          { type: "bytes32", name: "to", value: "BTC" },
+        ],
+      }
+    );
+
+    expect(logs).toHaveLength(1);
+    expect(logs).toEqual(
+      expect.arrayContaining([
+        {
+          level: "ERROR",
+          message: expect.stringMatching(
+            "templateId '0x50c604914d8ed35473149457a1a0912b785813b4e2e51bd2b75409ca25c50e1d' does not match expected templateId '[^']*'"
+          ),
+        },
+      ])
+    );
+    expect(Object.values(res)).toEqual([null]);
+    expect(spy).not.toHaveBeenCalled();
+  });
+
+  it("returns null if reserved parameter '_type' is missing", async () => {
+    const spy = jest.spyOn(adapter, "buildAndExecuteRequest") as any;
+
+    const oisesWithoutType = oises.map((o) => ({
+      ...o,
+      endpoints: o.endpoints.map((e) => ({
+        ...e,
+        reservedParameters: e.reservedParameters.filter(
+          (r) => r.name !== "_type"
+        ),
+      })),
+    }));
+
+    const [logs, res] = await readApiValue(
+      airnodeAddress,
+      oisesWithoutType,
+      apiCredentials,
+      job
+    );
+
+    expect(logs).toHaveLength(1);
+    expect(logs).toEqual(
+      expect.arrayContaining([
+        {
+          level: "ERROR",
+          message: expect.stringMatching(
+            "reserved parameter '_type' is missing for endpoint: convertToUSD"
+          ),
+        },
+      ])
+    );
+    expect(Object.values(res)).toEqual([null]);
+    expect(spy).not.toHaveBeenCalled();
+  });
+
+  it("returns null if reserved parameter '_type' is missing", async () => {
+    const spy = jest.spyOn(adapter, "buildAndExecuteRequest") as any;
+
+    const oisesWithoutType = oises.map((o) => ({
+      ...o,
+      endpoints: o.endpoints.map((e) => ({
+        ...e,
+        reservedParameters: e.reservedParameters.filter(
+          (r) => r.name !== "_type"
+        ),
+      })),
+    }));
+
+    const [logs, res] = await readApiValue(
+      airnodeAddress,
+      oisesWithoutType,
+      apiCredentials,
+      job
+    );
+
+    expect(logs).toHaveLength(1);
+    expect(logs).toEqual(
+      expect.arrayContaining([
+        {
+          level: "ERROR",
+          message: expect.stringMatching(
+            "reserved parameter '_type' is missing for endpoint: convertToUSD"
+          ),
+        },
+      ])
+    );
+    expect(Object.values(res)).toEqual([null]);
+    expect(spy).not.toHaveBeenCalled();
+  });
```

Comment:
What's the difference between this test and the previous one?

Reply:
The call-api.ts in Airkeeper is similar but not a copy paste from call-api.ts in Airnode and its only function receives different arguments. I know it might look redundant at first but since this is a different app I think it needs to have basic testing. This function might even change in the near future and I would be safer to refactor with tests in place.

Reply:
No, you misunderstood, I'm totally fine with having tests, it just seems that you have copied the same test twice, the `returns null if reserved parameter '_type' is missing` one.

Reply:
Right, sorry about that. It was not easy to see it in the diff. It seems I just duplicated that code accidentally. I'll remove it